src/utils.cpp

```diff
@@ -426,16 +426,7 @@ void utils::extract_filter(const std::string& line,
 	std::string& filter,
 	std::string& url)
 {
-	std::string::size_type pos = line.find_first_of(":", 0);
-	std::string::size_type pos1 = line.find_first_of(":", pos + 1);
-	filter = line.substr(pos + 1, pos1 - pos - 1);
-	pos = pos1;
-	url = line.substr(pos + 1, line.length() - pos);
-	LOG(Level::DEBUG,
-		"utils::extract_filter: %s -> filter: %s url: %s",
-		line,
-		filter,
-		url);
+	rs_extract_filter(line.c_str(), &filter[0], &url[0]);
```

Comment:
If `filter` and `url` are freshly created objects, their internal arrays are empty, so taking a pointer to first element would yield a past-the-end pointer. Writing to it will corrupt memory.

Do you see any other ways to implement this?

Reply:
Not sure, this is the only way I know as of now. My C++ skills are not that good, maybe I just wrote like 100 lines of C++ so far?

But from what I saw here, aren't we changing the pointer to the memory created by rust? Isn't pointing to the first byte enough?

Reply:
On the C++ side, this takes a pointer to the first element. On the Rust side, the code dereferences that pointer, and copies something in there. In other words, it's copying just one character, not the whole string.

What's worse, the C++ side might not have allocated any memory, so the pointer is *invalid*. Copying anything into it means corrupting memory. That's very, very bad thing, because it affects the program in unexpected ways: it might work and pass tests on my machine, leak memory and fail the tests on CI, and lead to kernel panic on some user's computer. That's one of the reasons we're moving to Rust: it helps avoid this sort of thing.

The best implementation I can think of is:
1. `rs_extract_filter` returns two pointers to `char*`, which are stored into local variables. Rnestler advised to use a struct to hold the pointers, and I agree; amongst other things, it'll let you avoid scary pointers-to-pointers (`char**`);
2. each pointer is wrapped into `RustString`, which is then assigned to `filter` or `url`.

`RustString` takes care of converting `char*` into `std::string`, and freeing memory afterwards.

Do you think you can implement that?

Reply:
> rs_extract_filter returns two pointers to char*, which are stored into local variables. Rnestler advised to use a struct to hold the pointers, and I agree; amongst other things, it'll let you avoid scary pointers-to-pointers (char**);

I was thinking about that but I am not sure how to return two pointers for `char*`.

> each pointer is wrapped into RustString, which is then assigned to filter or url.

Maybe I will check on that.

I can probably do it if I have more info.

Reply:
@Minoru Ah, I think I did it.